#763 wiki update for removal step (#833)
Azure/ResourceModules: docs/wiki/ModulesDesign.md

- Likewise, the `name` parameter we have in most modules should give some indication of the file it was deployed with. For example, a `min.parameters.json` parameter file for the virtual network module may have a `name` property with the value `sxx-az-vnet-min-001` where `min` relates to the prefix of the parameter file itself.
- A module should have as many parameter files as it needs to evaluate all parts of the module's functionality.
- Sensitive data should not be stored inside the parameter file but rather be injected by the use of [tokens](./ParameterFileTokens.md) or via a [key vault reference](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/key-vault-parameter?tabs=azure-cli#reference-secrets-with-static-id).

> # Resource Removal

Reviewer: @simonbMS , do you think it makes sense to move it to the [removal section](https://github.com/Azure/ResourceModules/wiki/PipelinesDesign#removal) of the pipeline design page?

There we have currently a brief decription we could extend with this content. I somewhat feel the removal is more pipeline/platform rather than Module-Design specific (as that focuses more on how a module should look like)

Author: Honestly I don't think it fits perfectly in neither place.
At the moment I moved it in the pipeline section, as the removal step was already mentioned there.

However, I think we should brainstorm a bit (outside of this PR 😄) on how to restructure the entire Wiki (this could even be one of the "outcome" of the hackathon).
Personally I would prefer a "user story based" structure (use/contribute), even if it could end up duplicating some content

Reviewer: Sounds good to me

Removal step is triggered after the deployment completes. This is used for several reasons:
- Make sure to keep the validation subscription cost as low as possible.
- Enable testing of the full module deployment at every run.

> The default removal procedure works fine for most of the modules created so far, so it's likely you won't have to change anything to make your module to be removed correctly after deployment.

Reviewer: ```suggestion
The default removal procedure works fine for most of the modules created so far, so it's likely you won't have to change anything to make a new module to be removed correctly after deployment.
```

Author: I will change it to "the module you're editing"

## How it works

The removal process will remove all resources created during deployment. The list is identified by:

> 1. Getting the list of resources created by your deployment (even if created by a "nested" deployment - a deployment created by the parent one -)

Reviewer: ```suggestion
1. Recursively fetch the list of resource IDs created through your deployment 
```

Author: Taken and added something to better explain

> 1. Ordering the list based on resourceId tokens length (ensures child resources are removed first)

Reviewer: ```suggestion
1. Ordering the list decending based on the resource IDs segment count (ensures child resources are removed first). E.g. `storageAccount/blobServices` comes before `storageAccount` as it has one more segmet (`/`).
```

Author: Taken and slightly modified

> 1. Removing from the list, if any, the resources used as dependencies for different modules (e.g. the commonly used Log Analytics workspace)

Reviewer: ```suggestion
1. Filtering any resources used as dependencies for different modules (e.g. the commonly used Log Analytics workspace) from the list
```

Author: Slightly modified

> 1. If provided, putting on the top of the list of resources to remove specific resource types (as resource id length may not in all cases)

Reviewer: ```suggestion
1. Moving specific resource types to the top of the list (if a certain order is required). For example `vWAN` requires its `Virtual Hubs` to be removed first, even though they are no child-resources.
```

Author: Modified as suggested

> After the removal of each resource, the script will attempt a **post removal operation**. This can be used for those resource types that requires a post processing, like purging a key vault.

Reviewer: ```suggestion
After a resource is removed, the script will optionally execute a **post removal operation**. This can be used for those resource types that requires a post processing, like purging a soft-deleted key vault.
```

Author: Taken almost as suggested

The procedure is initiated by the script `/utilities/pipelines/resourceRemoval/Initialize-DeploymentRemoval.ps1`, run during deployment by:
- (Azure DevOps) `/.azuredevops/pipelineTemplates/module.jobs.deploy.yml`
- (GitHub) `/.github/actions/templates/validateModuleDeployment/action.yml`

> It uses several helper scripts that can be found in `/utilities/pipelines/resourceRemoval`

Reviewer: ```suggestion
It uses several helper scripts that can be found in the `/utilities/pipelines/resourceRemoval/helper` folder
```

## Create a specialized removal procedure

> You can define a custom removal procedure by **modifying the resource types removal sequence** and by **defining a custom post-removal action**. The two are independent and you can act on one or both.

Reviewer: ```suggestion
You can define a custom removal procedure by **modifying the resource types removal sequence**, **defining a custom removal action** and by **defining a custom post-removal action**. They are independent and you can adapt them as needed.
```

Author: Added the third option

To modify the resource types removal sequence:
1. Open the `/utilities/pipelines/resourceRemoval/Initialize-DeploymentRemoval.ps1` file.
1. Look for the following comment: `### CODE LOCATION: Add custom removal sequence here`
1. Add a case value that matches your module name
1. In the case block, update the `$removalSequence` variable value to accommodate your module requirements
1. Remember to add the `break` statement.

> 

Reviewer: Add:

To add a custom removal step:
1. Open the `/utilities/pipelines/resourceRemoval/helper/Invoke-ResourceRemoval.ps1` file.
1. Look for the following comment: `### CODE LOCATION: Add custom removal sequence here`
1.  Add a case value that matches the resource type you want to add a custom removal operation for.

Author: @MrMCake  I totaly missed this method :) . Added it in my latest commit, please review
